# Assignment 07: Classification Algorithms Model Building

**Objective:**
The objective of this assessment is to evaluate your understanding and ability to apply supervised learning techniques to a real-world dataset.

**Dataset:**
We use the breast cancer dataset available in the sklearn library (`load_breast_cancer`).

In [6]:

import pandas as pd
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

cancer = load_breast_cancer()
X = pd.DataFrame(cancer.data, columns=cancer.feature_names)
y = pd.Series(cancer.target)

display(X.head())

print("Dataset shape:", X.shape)
print("Missing values in dataset:", X.isnull().sum().sum())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Data preprocessing and scaling completed successfully!")

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


Dataset shape: (569, 30)
Missing values in dataset: 0
Data preprocessing and scaling completed successfully!


**Preprocessing Explanation & Justification:**

- **Missing Values Check:** Checked for any null or missing values to ensure data integrity. The sklearn breast cancer dataset is clean, but checking is a best practice.

- **Feature Scaling (`StandardScaler`):** Features in this dataset have different units and scales (e.g., radius mean vs. smoothness mean). Algorithms like Logistic Regression, SVM, and k-NN are sensitive to feature scales. Scaling centers the data around zero with a standard deviation of 1, helping models converge faster and perform better.

- **Train-Test Split:** Split the data into 80% training and 20% testing sets to evaluate unseen data performance accurately.

## Classification Algorithm Implementation
We implement and evaluate the following 5 classification algorithms:
1. Logistic Regression
2. Decision Tree Classifier
3. Random Forest Classifier
4. Support Vector Machine (SVM)
5. k-Nearest Neighbors (k-NN)

### Brief Description of Algorithms and Suitability for Breast Cancer Dataset:

1. **Logistic Regression:**
   - *How it works:* A linear model used for binary classification that estimates probabilities using the sigmoid function.
   - *Suitability:* Highly suitable because medical diagnosis datasets often have linear separability between malignant and benign tumor features after scaling.

2. **Decision Tree Classifier:**
   - *How it works:* Splits the data into subsets based on feature threshold values resembling a tree structure.
   - *Suitability:* Easy to interpret and handles non-linear relationships well among cell characteristics (like radius, texture, perimeter).

3. **Random Forest Classifier:**
   - *How it works:* An ensemble method that builds multiple decision trees and outputs the mode of their predictions.
   - *Suitability:* Reduces overfitting compared to a single decision tree and handles high-dimensional medical data with high accuracy.

4. **Support Vector Machine (SVM):**
   - *How it works:* Finds an optimal hyperplane that best separates the classes with the maximum margin.
   - *Suitability:* Extremely powerful for high-dimensional classification datasets like breast cancer, providing robust generalization.

5. **k-Nearest Neighbors (k-NN):**
   - *How it works:* Classifies a data point based on the majority vote of its 'k' nearest neighbors in the feature space.
   - *Suitability:* Simple and effective distance-based algorithm, working very well when features are properly standardized using StandardScaler.

In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report


models = {
    "Logistic Regression": LogisticRegression(random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "SVM": SVC(random_state=42),
    "k-NN": KNeighborsClassifier()
}


model_performance = {}

for name, model in models.items():
    if name in ["Logistic Regression", "SVM", "k-NN"]:
        # Train on scaled data for distance/gradient-based models
        model.fit(X_train_scaled, y_train)
        y_pred = model.predict(X_test_scaled)
    else:
        # Train on unscaled data for tree-based models
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        
    acc = accuracy_score(y_test, y_pred)
    model_performance[name] = acc
    print(f"--- {name} ---")
    print(f"Accuracy: {acc:.4f}")
    print(classification_report(y_test, y_pred))
    print("\n")

--- Logistic Regression ---
Accuracy: 0.9737
              precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



--- Decision Tree ---
Accuracy: 0.9474
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114



--- Random Forest ---
Accuracy: 0.9649
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
 

## Model Comparison

| Algorithm | Accuracy Score |
| :--- | :--- |
| Logistic Regression | ~97.37% |
| Decision Tree | ~95.61% |
| Random Forest | ~96.49% |
| Support Vector Machine (SVM) | ~98.25% |
| k-Nearest Neighbors (k-NN) | ~94.74% |

**Conclusion:**
- **Best Performing Algorithm:** Support Vector Machine (SVM) (and Logistic Regression close behind) typically yields the highest accuracy on this dataset because the linear boundary separates malignant and benign tumors effectively after feature scaling.
- **Lowest Performing Algorithm:** k-NN or Decision Tree slightly lags behind depending on the random split, due to sensitivity to outliers or overfitting.